### PR TITLE
Adding a `min_steps` parameter to `BeamSearch`

### PR DESCRIPTION
**Pull request by danieldeutsch**
Implements feature request 1 from #5205.

Changes proposed in this pull request:

- Adds a `min_steps` parameter to `BeamSearch`, which allows specifying the minimum length of the predicted sequences. This is commonly used in tasks like summarization, for instance [by BART](https://github.com/pytorch/fairseq/blob/425c36eafff535fe7337f8bdd5ace22ebacc78cb/examples/bart/summarize.py#L10-L11)

## Before submitting

- [x] I've read and followed all steps in the [Making a pull request](https://github.com/allenai/allennlp/blob/main/CONTRIBUTING.md#making-a-pull-request)
    section of the `CONTRIBUTING` docs.
- [x] I've updated or added any relevant docstrings following the syntax described in the
    [Writing docstrings](https://github.com/allenai/allennlp/blob/main/CONTRIBUTING.md#writing-docstrings) section of the `CONTRIBUTING` docs.
- [ ] If this PR fixes a bug, I've added a test that will fail without my fix.
- [x] If this PR adds a new feature, I've added tests that sufficiently cover my new functionality.

## After submitting

- [ ] All GitHub Actions jobs for my pull request have passed.
- [ ] **`codecov/patch`** reports high test coverage (at least 90%).
    You can find this under the "Actions" tab of the pull request once the other checks have finished.

### CODE REVIEW

File: allennlp/nn/beam_search.py [hunk 1]
```diff
@@ -471,19 +475,25 @@ def __init__(
         beam_size: int = 10,
         per_node_beam_size: int = None,
         sampler: Sampler = None,
+        min_steps: int = 0,
```
Comment (epwalsh) [inline allennlp/nn/beam_search.py:478]: I think making this an `Optional[int]` is more clear.

```suggestion
        min_steps: Optional[int] = None,
```

**Review by epwalsh — COMMENTED** [review]
Thanks @danieldeutsch, this looks great! I just have one comment

File: allennlp/nn/beam_search.py (hunk 1 above)
Comment (danieldeutsch) [reply allennlp/nn/beam_search.py]: Ok, I pushed this change. I had to change the code to make sure the value is valid, too.

**Review by epwalsh — APPROVED** [review]
LGTM!

**Pull request merged**